lib/annon_api/plugins/proxy/adapters/http.ex

```diff
@@ -4,58 +4,126 @@ defmodule Annon.Plugins.Proxy.Adapters.HTTP do
   """
   alias Annon.Plugin.UpstreamRequest
   alias Plug.Conn
+  require Logger
 
   @drop_headers Application.get_env(:annon_api, :dropped_upstream_headers, [])
 
-  def dispatch(%UpstreamRequest{} = upstream_request, %Conn{method: method} = conn) do
-    upstream_url = UpstreamRequest.to_upstream_url!(upstream_request)
+  @stream_opts [
+    length: :infinity,      # Read whole response
+    read_length: 1_049_600, # in 1 Mb chunks
+    read_timeout: 15_000    # with 15 seconds timeout between chunks
+  ]
 
-    case Conn.get_req_header(conn, "content-type") do
-      [content_type | _] ->
-        if String.starts_with?(content_type, "multipart/form-data") do
-          do_fileupload_request_cont(upstream_url, upstream_request, conn, method)
-        else
-          do_request_cont(upstream_url, upstream_request, conn, method)
-        end
-      _ ->
-        do_request_cont(upstream_url, upstream_request, conn, method)
-    end
-  end
+  @buffer_opts [
+    connect_timeout: 30_000,
+    recv_timeout: 30_000,
+    timeout: 30_000
+  ]
+
+  def dispatch(%UpstreamRequest{} = upstream_request, %Conn{body_params: %Plug.Conn.Unfetched{}} = conn) do
+    upstream_url = UpstreamRequest.to_upstream_url!(upstream_request)
 
-  defp do_fileupload_request_cont(upstream_url, upstream_request, conn, _method) do
     req_headers =
       Enum.reject(upstream_request.headers, fn {k, _} ->
-        String.downcase(k) in ["content-type", "content-disposition", "content-length"] ++ @drop_headers
+        String.downcase(k) in @drop_headers
       end)
 
-    multipart = Annon.Plugins.Proxy.MultipartForm.reconstruct_using(conn.body_params)
+    method = String.to_atom(conn.method)
 
-    HTTPoison.post!(upstream_url, {:multipart, multipart}, req_headers)
+    with {:ok, client_ref} <- :hackney.request(method, upstream_url, req_headers, :stream, []),
+         {:ok, client_ref, conn} <- stream_request_body(client_ref, upstream_request, conn),
+         %Conn{} = conn <- stream_response_body(client_ref, conn) do
+      {:ok, conn}
+    else
+      {:error, term} ->
+        Conn.send_resp(conn, 502, Annon.Helpers.Response.build_upstream_error(to_string(term)))
+    end
   end
 
-  defp do_request_cont(upstream_url, upstream_request, conn, method) do
+  def dispatch(%UpstreamRequest{} = upstream_request, %Conn{method: method} = conn) do
+    upstream_url = UpstreamRequest.to_upstream_url!(upstream_request)
+
+    # TODO: find better solution that dropping this for broken HTTPBin
```

Comment:
Found a TODO tag in a comment: # TODO: find better solution that dropping this for broken HTTPBin